src/workflow/workflow_tasks/mod.rs

```diff
@@ -425,55 +426,65 @@ impl WorkflowTaskManager {
                 }
             }
 
-            let (are_pending, server_cmds, local_activities) = machine_mut!(
+            let (are_pending, server_cmds, local_activities, wft_timeout) = machine_mut!(
                 self,
                 run_id,
                 Some(task_token.clone()),
-                |wfm: &mut WorkflowManager| async move {
-                    // Send commands from lang into the machines then check if the workflow run needs
-                    // another activation and mark it if so
-                    wfm.push_commands(commands).await?;
-                    let are_pending = wfm.apply_next_task_if_ready().await?;
-                    // We want to fetch the outgoing commands only after a next WFT may have been
-                    // applied, as outgoing server commands may be affected.
-                    Ok((
-                        are_pending,
-                        wfm.get_server_commands(),
-                        wfm.drain_queued_local_activities(),
-                    ))
+                |wfm: &mut WorkflowManager| {
+                    async move {
+                        // Send commands from lang into the machines then check if the workflow run
+                        // needs another activation and mark it if so
+                        wfm.push_commands(commands).await?;
+                        let are_pending = wfm.apply_next_task_if_ready().await?;
+                        // We want to fetch the outgoing commands only after a next WFT may have
+                        // been applied, as outgoing server commands may be affected.
+                        let outgoing_cmds = wfm.get_server_commands();
+                        let new_local_acts = wfm.drain_queued_local_activities();
+
+                        let wft_timeout: Duration = wfm
+                            .machines
+                            .started_attrs()
+                            .and_then(|attrs| attrs.workflow_task_timeout.clone())
+                            .and_then(|tt| tt.try_into().ok())
+                            .ok_or_else(|| {
+                                WFMachinesError::Fatal(
+                                    "Workflow's start attribs were missing a well formed task timeout"
+                                        .to_string(),
+                                )
+                            })?;
+
+                        Ok((are_pending, outgoing_cmds, new_local_acts, wft_timeout))
+                    }
+                    .boxed()
                 }
-                .boxed()
             )?;
 
             if are_pending {
                 self.needs_activation(run_id);
             }
+            local_activity_request_sink(local_activities);
+
+            // The heartbeat deadline is 80% of the WFT timeout
+            let wft_heartbeat_deadline = start_time.add(wft_timeout.mul_f32(0.8));
```

Comment:
Can we have a constant that defines this?
Would be nice if this was easier to find.